stratum/hal/lib/common/hal.cc

```diff
@@ -253,6 +257,8 @@ ::util::Status Hal::Run() {
 
 void Hal::HandleSignal(int value) {
   LOG(INFO) << "Received signal: " << strsignal(value);
+  // Need to do Teardown first to clear up any services first
+  Teardown();
```

Comment:
Yes I added this Teardown in the signal handler as interrupts were hanging and not cleaning up (i.e. the Run Wait() was waiting forever.  However, if process dies without going thru signal handler then we'll still need to call Teardown so was leaving both in?

Reply:
Looking at the code some more, it should be fine to get rid of the Teardown call in the Run() function as the Signal Handler should be called in all circumstances to shutdown the server.

Reply:
The `Wait()` should return after we called `Shutdown()`(https://github.com/stratum/stratum/pull/43/files#diff-4a43a81e5c12f8b203e802dcf2e15857R268)

I just realized that we still need to put `Teardown()` call to `Run` function because we need to get status from `Teardown()` and show error messages in main function, for example:
https://github.com/stratum/stratum/blob/master/stratum/hal/bin/barefoot/main.cc#L189
https://github.com/stratum/stratum/blob/master/stratum/hal/bin/bcm/standalone/main.cc#L135
https://github.com/stratum/stratum/blob/master/stratum/hal/bin/bmv2/main.cc#L184

Reply:
I'm happy to pull this change out from my PR if you want to just leave it the way it was, but I found that the Wait was hanging when there is an outstanding Subscribe request so put the Teardown in the Signal Handler and left it in the Run function as well just to be sure.

Reply:
ok not a bad idea Max.  I think that the reason I put it in here is that for the most part interrupting the stratum binary cleans up properly unless you have an outstanding Subscribe request (which issues a read on a channel).  The comments around the Shutdown call using now() to interrupt active Reads doesn't seem to work for Reads on a channel from memory.

Reply:
do you want to try and pull these two commits or do you want me to make another to pull it out?

Reply:
Whatever is easier for you. In the end, we squash all commits anyway.
Just make sure to keep the changes somewhere to submit them in another PR.

Reply:
I've created another commit to remove this change.